citest/gcp_testing/google_idp_session.py

```diff
@@ -0,0 +1,195 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Standard python modules.
+import requests
+import urllib2
+
+from bs4 import BeautifulSoup
+
+# Our modules.
+from ..service_testing import request_mutator
+from ..service_testing import idp_manager
+
+GOOGLE_ACCOUNT_LOGIN_URL = 'https://accounts.google.com/ServiceLogin'
+GOOGLE_ACCOUNT_LOGOUT_URL = 'https://www.google.com/accounts/Logout'
+
+class GoogleIDPSession(request_mutator.RequestMutator,
+                       idp_manager.IDPSessionManager):
+  """Utility for accessing applications that use Google as an identity provider.
+
+  This automates Google OAuth2/SAML account authentication and scope approval
+  after a redirect from a third party application.
+
+  The intended workflow is the following:
+
+  * Target some resource in a third party application.
+  * The third party application needs some information in your Google account.
+  * The third party application redirects you to Google to sign in and approve
+    the scopes the application is requesting. (This is automated by this class).
+  * A Http session is established with Google and tracked with Cookies in
+    the request/response headers.
+  * Possibly, a Http session is established with the third party application
+    as well.
+  * Make requests in the established session scope.
+  * Log out after the necessary exchange is made.
+
+  This class automtes both the Google account login and scope approval to create
+  an authenticated HTTP session, and provides access to the session cookies for
+  use with urllib2 Requests. Both direct access to the Cookies and mutating
+  requests via RequestMutator methods are supported.
+
+  Example usage:
+    # target_url should forward to 'accounts.google.com/ServiceLogin' page
+    # via an OAuth2 or SAML redirect.
+    target_url = "http://<gate_host>/credentials"
+    logout_url = "http://<gate_host>/auth/logout"
+    session = GoogleIDPSession(target_url,
+                           "user@domain.net",
+                           "secret_pass_word",
+                           logout_url=logout_url)
+    session.initialize()
+    req = urllib2.Request(url=target_url)
+    self.mutate_request(req) # loads cookies into request header
+    resp = urllib2.urlopen(req)
+    session.terminate()
+  """
+
+
+  def __init__(
+      self, target_url, login, password,
+      user_approval_url='https://accounts.google.com/ServiceLoginAuth',
+      logout_url=None):
+    """Construct an instance.
+
+    Args:
+      target_url [String]: A URL that will eventually redirect to
+        'https://accounts.google.com/ServiceLogin' given a new session.
+      login [String]: Username/email to login as.
+      password: [String] Password for the provided login.
```

Comment:
Where do login/password come from? Specifically, does Google know about them? How? Give a URL to the google docs. Probably state this up front in the class header. Prerequisites to the flow you describe above.